File: src/api/local-shell.ts
import { invokeCommand } from '@/api/base'

/** 创建本地 Shell 会话 */
export function localShellSpawn(sessionId: string, cols: number, rows: number): Promise<void> {
  return invokeCommand('local_shell_spawn', { sessionId, cols, rows })
}

/** 向本地 Shell 发送输入 */
export function localShellWrite(sessionId: string, data: string): Promise<void> {
  return invokeCommand('local_shell_write', { sessionId, data })
}

/** 调整本地 Shell 终端大小 */
export function localShellResize(sessionId: string, cols: number, rows: number): Promise<void> {
  return invokeCommand('local_shell_resize', { sessionId, cols, rows })
}

/** 关闭本地 Shell 会话 */
export function localShellClose(sessionId: string): Promise<void> {
  return invokeCommand('local_shell_close', { sessionId })
}

File: src-tauri/src/commands/local_shell.rs
use std::sync::Arc;

use tauri::Manager;

use crate::services::local_shell_engine::LocalShellEngine;
use crate::utils::error::AppError;

/// 内部使用 RwLock，无需外层 Mutex
pub type LocalShellEngineState = Arc<LocalShellEngine>;

#[tauri::command]
pub async fn local_shell_spawn(
    app: tauri::AppHandle,
    session_id: String,
    cols: u32,
    rows: u32,
) -> Result<(), AppError> {
    let engine = app.state::<LocalShellEngineState>().inner().clone();
    engine
        .spawn_shell(&app, &session_id, cols as u16, rows as u16)
        .await
}

#[tauri::command]
pub async fn local_shell_write(
    app: tauri::AppHandle,
    session_id: String,
    data: String,
) -> Result<(), AppError> {
    let engine = app.state::<LocalShellEngineState>().inner().clone();
    engine.write_data(&session_id, data.as_bytes()).await
}

#[tauri::command]
pub async fn local_shell_resize(
    app: tauri::AppHandle,
    session_id: String,
    cols: u32,
    rows: u32,
) -> Result<(), AppError> {
    let engine = app.state::<LocalShellEngineState>().inner().clone();
    engine.resize(&session_id, cols as u16, rows as u16).await
}

#[tauri::command]
pub async fn local_shell_close(
    app: tauri::AppHandle,
    session_id: String,
) -> Result<(), AppError> {
    let engine = app.state::<LocalShellEngineState>().inner().clone();
    engine.close_shell(&session_id).await
}
